repose-aggregator/components/filters/keystone-v2-filter/src/main/scala/org/openrepose/filters/keystonev2/KeystoneV2Authorization.scala

```diff
@@ -38,116 +38,118 @@ object KeystoneV2Authorization extends LazyLogging {
+        .map(getRequestTenants(_, request))
+        .getOrElse(Set.empty)
+      val tenantPrefixes = validateTenant
+        .map(_.getStripTokenTenantPrefixes)
+        .flatMap(Option.apply)
+        .map(_.split('/'))
+        .getOrElse(Array.empty)
+      val scopedTenantToRolesMap = validateTenant
+        .map(_ => getScopedTenantToRolesMap(tenantPrefixes, userTenantToRolesMap, requestTenants))
+      val matchedTenants = (scopedTenantToRolesMap.getOrElse(Map.empty) - DomainRoleTenantKey).keySet
+      val tenantAuthorization = validateTenant
+        .map(_ => authorizeTenant(tenantPrefixes, matchedTenants, requestTenants))
+        .getOrElse(Success(Unit))
+
+      val endpointAuthorization = Option(config.getRequireServiceEndpoint)
+        .map(requiredEndpoint => authorizeEndpoints(requiredEndpoint, endpoints))
+        .getOrElse(Success(Unit))
+
+      tenantAuthorization.flatMap(_ => endpointAuthorization) match {
+        case Success(_) => AuthorizationPassed(scopedTenantToRolesMap.getOrElse(userTenantToRolesMap), matchedTenants)
+        case Failure(e) => AuthorizationFailed(scopedTenantToRolesMap.getOrElse(userTenantToRolesMap), matchedTenants, e)
+      }
     }
   }
 
-  // TODO: Switch to getRequestTenants (plural) and include all tenants extracted. This is a behavior change.
-  def getRequestTenant(config: ValidateTenantType, request: HttpServletRequestWrapper): String = {
-    Option(config).flatMap(validateTenantConfig =>
-      (validateTenantConfig.getUriExtractionRegexAndHeaderExtractionName.asScala.toStream map {
-        case headerExtraction: HeaderExtractionType =>
-          request.getPreferredSplittableHeaders(headerExtraction.getValue).asScala.headOption
-        case uriExtraction: UriExtractionType =>
-          val uriRegex = uriExtraction.getValue.r
-          request.getRequestURI match {
-            case uriRegex(tenantId, _*) => Some(tenantId)
-            case _ => None
-          }
-        case _ =>
-          logger.error("An unexpected tenant extraction type was encountered")
-          None
-      }).find(_.nonEmpty).flatten
-    ).getOrElse(throw UnparsableTenantException("Could not parse tenant from the URI and/or the configured header"))
+  // TODO: Rather than throwing an exception, this method should return a Try.
+  // NOTE: Throwing an exception currently works because this method is always called in a [[Try]] stack.
+  @throws[UnparsableTenantException]
+  def getRequestTenants(config: ValidateTenantType, request: HttpServletRequestWrapper): Set[String] = {
+    config.getUriExtractionRegexAndHeaderExtractionName.asScala.toSet[ExtractionType] flatMap {
+      case headerExtraction: HeaderExtractionType =>
+        request.getSplittableHeaderScala(headerExtraction.getValue)
+      case uriExtraction: UriExtractionType =>
+        val uriRegex = uriExtraction.getValue.r
+        request.getRequestURI match {
+          case uriRegex(all @ _*) => all
```

Comment:
I want to point out the change here. Instead of only extracting the value from the first capture group, this will now extract all values from all capturing capture groups, and all extracted values must be validated. I am slightly concerned that this is not backwards compatible with current configurations.

Of course, you can make a capturing group non-capturing using the prefix `?:` like so: `(?:hello)`.

Reply:
So that's a valid concern. I'd say revert the behavior back to what it use to be, except the uri extractor filter also only works on the first capture group. We seem to assume that things will be the first capture group in several places. I think I lean towards reverting this behavior back to what it use to be. Deprecating the uri extraction config here if we haven't already, and adding a fast follow story to make uri extract to header filter work on every capture group. We also have a potential backwards compatibility problem there as well, but we can finagle the config if we need to easier there.

Reply:
Will do.